dali/kernels/slice/slice_gpu.cuh

```diff
@@ -36,78 +37,189 @@ template <int Dims>
 struct SliceSampleDesc {
   void *__restrict__ out;
   const void *__restrict__ in;
+
+  const void *__restrict__ fill_values;
+  int channel_dim;
+  bool need_pad;
+
   TensorShape<Dims> in_strides;
-  TensorShape<Dims> out_strides;
+  fast_div<uint64_t> out_strides[Dims];
+
+  TensorShape<Dims> anchor;
+  TensorShape<Dims> in_shape;
 };
 
 struct BlockDesc {
   int sampleIdx;
-  size_t offset;
-  size_t size;
+  uint64_t offset;
+  uint64_t size;
 };
 
+/**
+ * @brief Simplified algorithm when no padding is necessary
+ * @remarks `in` already refers to the slice anchor start
+ */
 template <int Dims, typename OutputType, typename InputType>
-__device__ void SliceFunc(OutputType *__restrict__ out, const InputType *__restrict__ in,
-                          const int64_t *out_strides, const int64_t *in_strides,
-                          size_t offset, size_t block_end) {
-  if (Dims > 1 && out_strides[Dims-1] == in_strides[Dims-1]) {
-    const int NextDims = Dims > 1 ? Dims-1 : 1;
-    SliceFunc<NextDims>(out, in, out_strides, in_strides, offset, block_end);
+__device__ void SliceFuncNoPad(OutputType *__restrict__ out, const InputType *__restrict__ in,
+                               const fast_div<uint64_t> *out_strides, const int64_t *in_strides,
+                               uint64_t offset, uint64_t block_end) {
+  if (Dims > 1 && out_strides[Dims - 1] == in_strides[Dims - 1]) {
+    const int NextDims = Dims > 1 ? Dims - 1 : 1;
+    SliceFuncNoPad<NextDims, OutputType, InputType>(out, in, out_strides, in_strides, offset,
+                                                    block_end);
     return;
   }
 
   for (; offset < block_end; offset += blockDim.x) {
-    size_t idx = offset;
-    size_t out_idx = idx;
-    size_t in_idx = 0;
+    uint64_t idx = offset;
+    uint64_t out_idx = idx;
+    uint64_t in_idx = 0;
     for (int d = 0; d < Dims; d++) {
-      int i_d = idx / out_strides[d];
-      idx %= static_cast<size_t>(out_strides[d]);
-      in_idx += i_d * static_cast<size_t>(in_strides[d]);
+      int i_d = div_mod(idx, idx, out_strides[d]);
```

Comment:
👍

Reply:
Do you know what is the perf diff, if any?

Reply:
not yet

Reply:
The Crop operator benchmarks we have are showing around 5% performance boost with this new implementation